src/metabase/query_processor/middleware/annotate.clj

```diff
@@ -480,23 +480,36 @@
                         (mbql-cols source-query results))]
     (qp.util/combine-metadata columns source-metadata)))
 
+(defn- add-aliases-to-legacy-cols [inner-query cols]
+  (let [returned (lib.metadata.calculation/returned-columns (mlv2-query inner-query))
+        aliases  (into {} (for [col returned]
+                            [[(:name col) (:metabase.lib.join/join-alias col)]
+                             (:lib/desired-column-alias col)]))]
+    (mapv (fn [col]
+            (let [k     [(:name col) (:source_alias col)]
+                  alias (get aliases k)]
+              (cond-> col
+                alias (assoc :desired_column_alias alias))))
```

Comment:
Isn't this the assoc-some pattern you are advocating?

Reply:
Yep, so it is.